File: aws/ses/typescript/src/http/transport.ts
/**
 * Transport layer abstraction for HTTP communication.
 *
 * This module provides a pluggable transport layer for sending HTTP requests.
 * The default implementation uses the Fetch API, which is available in Node.js 18+
 * and all modern browsers.
 *
 * @module http/transport
 */

import type { HttpRequest, HttpResponse, HttpClientConfig } from './types.js';

/**
 * Transport interface for HTTP communication.
 *
 * This abstraction allows for different HTTP implementations
 * (e.g., fetch, undici, mock for testing).
 *
 * @example
 * ```typescript
 * class CustomTransport implements Transport {
 *   async send(request: HttpRequest): Promise<HttpResponse> {
 *     // Custom implementation
 *     return { status: 200, headers: {}, body: '' };
 *   }
 * }
 * ```
 */
export interface Transport {
  /**
   * Send an HTTP request and return the response.
   *
   * @param request - The HTTP request to send
   * @returns Promise resolving to the HTTP response
   * @throws Error if the request cannot be sent or times out
   */
  send(request: HttpRequest): Promise<HttpResponse>;
}

/**
 * Fetch-based HTTP transport implementation.
 *
 * This transport uses the native Fetch API for HTTP communication.
 * It supports timeouts, connection management, and proper error handling.
 *
 * @example
 * ```typescript
 * const transport = new FetchTransport({
 *   timeout: 30000,
 *   connectTimeout: 10000,
 *   keepAlive: true
 * });
 *
 * const response = await transport.send({
 *   method: 'GET',
 *   url: 'https://api.example.com/data',
 *   headers: { 'Accept': 'application/json' }
 * });
 * ```
 */
export class FetchTransport implements Transport {
  private readonly config: Required<HttpClientConfig>;

  /**
   * Create a new fetch transport.
   *
   * @param config - Optional configuration for the transport
   *
   * @example
   * ```typescript
   * const transport = new FetchTransport({
   *   timeout: 60000,      // 60 second timeout
   *   connectTimeout: 5000, // 5 second connection timeout
   *   keepAlive: true      // Enable connection reuse
   * });
   * ```
   */
  constructor(config?: HttpClientConfig) {
    this.config = {
      timeout: config?.timeout ?? 30000,
      connectTimeout: config?.connectTimeout ?? 10000,
      keepAlive: config?.keepAlive ?? true,
    };
  }

  /**
   * Send an HTTP request using the Fetch API.
   *
   * This method handles:
   * - Request timeout via AbortController
   * - Proper header forwarding
   * - Response body reading
   * - Error conversion to standardized format
   *
   * @param request - The HTTP request to send
   * @returns Promise resolving to the HTTP response
   * @throws Error if the request fails or times out
   *
   * @example
   * ```typescript
   * const transport = new FetchTransport();
   *
   * try {
   *   const response = await transport.send({
   *     method: 'POST',
   *     url: 'https://api.example.com/data',
   *     headers: { 'Content-Type': 'application/json' },
   *     body: JSON.stringify({ key: 'value' })
   *   });
   *   console.log('Status:', response.status);
   * } catch (error) {
   *   console.error('Request failed:', error);
   * }
   * ```
   */
  async send(request: HttpRequest): Promise<HttpResponse> {
    // Create an AbortController for timeout handling
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      // Prepare fetch options
      const fetchOptions: RequestInit = {
        method: request.method,
        headers: request.headers,
        body: request.body,
        signal: controller.signal,
      };

      // Add keep-alive if enabled (Node.js specific)
      if (this.config.keepAlive && typeof (globalThis as any).process !== 'undefined') {
        // @ts-ignore - keepalive is a Node.js-specific option
        fetchOptions.keepalive = true;
      }

      // Execute the fetch request
      const response = await fetch(request.url, fetchOptions);

      // Read the response body
      const body = await response.text();

      // Extract headers
      const headers: Record<string, string> = {};
      response.headers.forEach((value, key) => {
        headers[key.toLowerCase()] = value;
      });

      return {
        status: response.status,
        headers,
        body,
      };
    } catch (error) {
      // Handle abort/timeout errors
      if (error instanceof Error && error.name === 'AbortError') {
        throw new Error(`Request timeout after ${this.config.timeout}ms`);
      }

      // Handle network errors
      if (error instanceof TypeError && error.message.includes('fetch')) {
        throw new Error(`Network error: ${error.message}`);
      }

      // Re-throw other errors
      throw error;
    } finally {
      // Clean up the timeout
      clearTimeout(timeoutId);
    }
  }

  /**
   * Get the current transport configuration.
   *
   * @returns The transport configuration
   *
   * @example
   * ```typescript
   * const transport = new FetchTransport({ timeout: 60000 });
   * const config = transport.getConfig();
   * console.log('Timeout:', config.timeout); // 60000
   * ```
   */
  getConfig(): Readonly<HttpClientConfig> {
    return { ...this.config };
  }
}

/**
 * Create a default transport instance.
 *
 * This is a convenience function for creating a FetchTransport
 * with default settings.
 *
 * @param config - Optional configuration
 * @returns A new FetchTransport instance
 *
 * @example
 * ```typescript
 * const transport = createDefaultTransport();
 *
 * // Or with custom config:
 * const customTransport = createDefaultTransport({
 *   timeout: 60000
 * });
 * ```
 */
export function createDefaultTransport(config?: HttpClientConfig): Transport {
  return new FetchTransport(config);
}

File: aws/ses/rust/src/http/transport.rs
//! Transport layer abstraction for HTTP communication.
//!
//! This module provides a pluggable transport layer for sending HTTP requests.
//! The default implementation uses reqwest, but other implementations can be
//! provided for testing or alternative HTTP backends.

use async_trait::async_trait;
use reqwest::{Client, Request, Response};
use std::sync::Arc;
use std::time::Duration;

use crate::error::{SesError, SesResult};
use super::pool::{ConnectionPool, PoolConfig};

/// Trait for HTTP transport implementations.
///
/// This trait abstracts the actual HTTP transport, allowing for different
/// implementations (e.g., reqwest, hyper, mock for testing).
#[async_trait]
pub trait Transport: Send + Sync {
    /// Send an HTTP request and return the response.
    ///
    /// # Arguments
    ///
    /// * `request` - The HTTP request to send
    ///
    /// # Returns
    ///
    /// The HTTP response or an error.
    ///
    /// # Errors
    ///
    /// Returns an error if the request cannot be sent or times out.
    async fn send(&self, request: Request) -> SesResult<Response>;

    /// Clone the transport into a new Arc.
    ///
    /// This is used to create new instances of the transport for use
    /// in different contexts while sharing the underlying connection pool.
    fn clone_box(&self) -> Arc<dyn Transport>;
}

/// Reqwest-based HTTP transport implementation.
///
/// This transport uses the reqwest library for HTTP communication and includes
/// connection pooling for efficient connection reuse.
pub struct ReqwestTransport {
    /// The reqwest HTTP client
    client: Client,

    /// Connection pool for managing connections
    pool: Arc<ConnectionPool>,
}

impl ReqwestTransport {
    /// Create a new reqwest transport with default settings.
    ///
    /// # Arguments
    ///
    /// * `timeout` - Request timeout duration
    /// * `connect_timeout` - Connection timeout duration
    ///
    /// # Returns
    ///
    /// A new `ReqwestTransport` instance or an error if client creation fails.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use std::time::Duration;
    /// use integrations_aws_ses::http::ReqwestTransport;
    ///
    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// let transport = ReqwestTransport::new(
    ///     Duration::from_secs(30),
    ///     Duration::from_secs(10)
    /// )?;
    /// # Ok(())
    /// # }
    /// ```
    pub fn new(timeout: Duration, connect_timeout: Duration) -> SesResult<Self> {
        let pool_config = PoolConfig::default();
        Self::with_pool_config(timeout, connect_timeout, pool_config)
    }

    /// Create a new reqwest transport with custom pool configuration.
    ///
    /// # Arguments
    ///
    /// * `timeout` - Request timeout duration
    /// * `connect_timeout` - Connection timeout duration
    /// * `pool_config` - Connection pool configuration
    ///
    /// # Returns
    ///
    /// A new `ReqwestTransport` instance or an error if client creation fails.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use std::time::Duration;
    /// use integrations_aws_ses::http::{ReqwestTransport, PoolConfig};
    ///
    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// let pool_config = PoolConfig {
    ///     max_idle_per_host: 10,
    ///     idle_timeout: Duration::from_secs(90),
    ///     max_lifetime: Some(Duration::from_secs(300)),
    /// };
    ///
    /// let transport = ReqwestTransport::with_pool_config(
    ///     Duration::from_secs(30),
    ///     Duration::from_secs(10),
    ///     pool_config
    /// )?;
    /// # Ok(())
    /// # }
    /// ```
    pub fn with_pool_config(
        timeout: Duration,
        connect_timeout: Duration,
        pool_config: PoolConfig,
    ) -> SesResult<Self> {
        let client = Client::builder()
            .timeout(timeout)
            .connect_timeout(connect_timeout)
            .pool_idle_timeout(pool_config.idle_timeout)
            .pool_max_idle_per_host(pool_config.max_idle_per_host)
            .tcp_keepalive(Some(Duration::from_secs(60)))
            .http2_keep_alive_interval(Some(Duration::from_secs(30)))
            .http2_keep_alive_timeout(Duration::from_secs(20))
            .http2_keep_alive_while_idle(true)
            .build()
            .map_err(|e| SesError::Transport {
                message: format!("Failed to create HTTP client: {}", e),
                source: Some(Box::new(e)),
                retryable: false,
            })?;

        let pool = Arc::new(ConnectionPool::new(pool_config));

        Ok(Self { client, pool })
    }

    /// Get a reference to the underlying reqwest client.
    ///
    /// This is useful for advanced use cases where direct access to the
    /// reqwest client is needed.
    pub fn client(&self) -> &Client {
        &self.client
    }

    /// Get a reference to the connection pool.
    ///
    /// This is useful for monitoring pool statistics or health.
    pub fn pool(&self) -> &ConnectionPool {
        &self.pool
    }
}

#[async_trait]
impl Transport for ReqwestTransport {
    async fn send(&self, request: Request) -> SesResult<Response> {
        // Acquire a connection slot from the pool
        self.pool.acquire().await?;

        // Execute the request
        let result = self.client.execute(request).await;

        // Release the connection slot back to the pool
        self.pool.release();

        // Convert reqwest error to SesError
        result.map_err(Into::into)
    }

    fn clone_box(&self) -> Arc<dyn Transport> {
        Arc::new(Self {
            client: self.client.clone(),
            pool: Arc::clone(&self.pool),
        })
    }
}

impl Clone for ReqwestTransport {
    fn clone(&self) -> Self {
        Self {
            client: self.client.clone(),
            pool: Arc::clone(&self.pool),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_reqwest_transport_creation() {
        let transport = ReqwestTransport::new(
            Duration::from_secs(30),
            Duration::from_secs(10),
        );
        assert!(transport.is_ok());
    }

    #[test]
    fn test_reqwest_transport_with_pool_config() {
        let pool_config = PoolConfig {
            max_idle_per_host: 5,
            idle_timeout: Duration::from_secs(60),
            max_lifetime: Some(Duration::from_secs(300)),
        };

        let transport = ReqwestTransport::with_pool_config(
            Duration::from_secs(30),
            Duration::from_secs(10),
            pool_config,
        );
        assert!(transport.is_ok());
    }

    #[test]
    fn test_reqwest_transport_clone() {
        let transport = ReqwestTransport::new(
            Duration::from_secs(30),
            Duration::from_secs(10),
        )
        .unwrap();

        let cloned = transport.clone();

        // Both should use the same pool (Arc::ptr_eq would be true)
        assert_eq!(
            Arc::as_ptr(&transport.pool),
            Arc::as_ptr(&cloned.pool)
        );
    }

    #[tokio::test]
    async fn test_transport_trait_object() {
        let transport = ReqwestTransport::new(
            Duration::from_secs(30),
            Duration::from_secs(10),
        )
        .unwrap();

        let transport_arc = transport.clone_box();

        // Should be able to use as trait object
        let _: &dyn Transport = &*transport_arc;
    }
}
